supabase/migrations/20230817110845_access_token.sql

```diff
@@ -0,0 +1,174 @@
+create table app.access_tokens(
+    id uuid primary key default gen_random_uuid(),
+    user_id uuid not null references auth.users(id),
+    token_hash bytea not null,
+    token_name text not null check (length(token_name) <= 64),
+    created_at timestamptz not null default now(),
+
+    unique (user_id, token_name)
+);
+
+grant insert
+    (id, user_id, token_hash, token_name)
+    on app.access_tokens
+    to authenticated;
+
+grant delete
+    on app.access_tokens
+    to authenticated;
+
+alter table app.access_tokens enable row level security;
+
+create policy access_tokens_select_policy
+    on app.access_tokens
+    as permissive
+    for select
+    to public
+    using ( auth.uid() = user_id );
+
+create policy access_tokens_insert_policy
+    on app.access_tokens
+    as permissive
+    for insert
+    to authenticated
+    with check ( auth.uid() = user_id );
+
+create policy access_tokens_delete_policy
+    on app.access_tokens
+    as permissive
+    for delete
+    to authenticated
+    using ( auth.uid() = user_id );
+
+create or replace function public.new_access_token(
+    token_name text
+)
+    returns text
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+    token bytea = gen_random_bytes(16);
+    token_hash bytea = pgsodium.crypto_pwhash_str(token);
+    token_text text = encode(token, 'hex');
+begin
+    begin
+        insert into app.access_tokens(user_id, token_hash, token_name)
+        values (account.id, token_hash, token_name);
+    exception when unique_violation then
+        raise exception 'Token with name `%s` already exists', token_name;
+    end;
```

Comment:
This will quite literally never happen, so I would omit it and panic when you see the insert violation, as the token being inserted will never repeat while this universe exists.

Reply:
Ah sorry, it's about the name. It's fine, then though I don't see the point in enforcing name uniqueness -- it's just for the UX. By making the token's prefix available in the database, debugging will be easier for people.

Reply:
I have removed the unique constraint on (token_name, user_id).